subsys/bluetooth/host/settings.c

```diff
@@ -19,21 +19,63 @@
 #include "hci_core.h"
 #include "settings.h"
 
+u8_t bt_settings_u8_to_dec(char *buf, u8_t buflen, u8_t value)
+{
+        u8_t divisor = 100;
+	u8_t num_digits = 0;
+	u8_t digit;
+
+	while (buflen > 0 && divisor > 0) {
+		digit = value / divisor;
+		if (digit != 0 || divisor == 1 || num_digits != 0) {
+			*buf = (char)digit + '0';
+			buf++;
+			buflen--;
+			num_digits++;
+		}
+		value -= digit * divisor;
+		divisor /= 10;
+	}
+	if (buflen) {
+		*buf = 0;
+	}
+	return num_digits;
+}
+
 void bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
 			    bt_addr_le_t *addr, const char *key)
 {
-	if (key) {
-		snprintk(path, path_size,
-			 "bt/%s/%02x%02x%02x%02x%02x%02x%u/%s", subsys,
-			 addr->a.val[5], addr->a.val[4], addr->a.val[3],
-			 addr->a.val[2], addr->a.val[1], addr->a.val[0],
-			 addr->type, key);
-	} else {
-		snprintk(path, path_size,
-			 "bt/%s/%02x%02x%02x%02x%02x%02x%u", subsys,
-			 addr->a.val[5], addr->a.val[4], addr->a.val[3],
-			 addr->a.val[2], addr->a.val[1], addr->a.val[0],
-			 addr->type);
+	size_t len = 3;
+
+	/* Key format:
+	     "bt/<subsys>/<addr><type>/<key>", "/<key>" is optional */
+	/* Skip if path_size is less than 3; strlen("bt/") */
+	if (len < path_size) {
+		strncpy(path, "bt/", path_size);
+		strncpy(path + len, subsys, path_size - len);
+		len = strlen(path);
+		if (len < path_size) {
+			strncpy(path + len, "/", path_size - len);
+			len++;
+		}
+		for (s8_t i = 5; i >= 0 && len < path_size; i--) {
+			len += bin2hex(&addr->a.val[i], 1, path + len,
+					      path_size - len);
+		}
+		len += bt_settings_u8_to_dec(
+			path + len, path_size - len, addr->type);
+		if (key && len < path_size) {
+			strncpy(path + len, "/", path_size - len);
+			len++;
+			strncpy(path + len, key, path_size - len);
+			len += strlen(path + len);
+		}
+		if (len >= path_size) {
+			/* Truncate string */
+			path[path_size - 1] = 0;
```

Comment:
`0` is an integer, `'\0'` is the null terminator character, please use the latter so it's clear what the target type is